src/kernel.ts

```diff
@@ -1,1069 +1,802 @@
+  }
+
+  private _id: string;
+  private _name: string;
+  private _kernelServiceUrl: string;
+  private _kernelUrl: string;
+  private _wsUrl: string;
+  private _username: string;
+  private _sessionId: string;
+  private _ws: WebSocket;
+  private _infoReply: any;
+  private _WebSocket: any;
+  private _reconnectLimit: number;
+  private _autorestartAttempt: number;
+  private _reconnectAttempt: number;
+  private _handlerMap: Map<string, KernelFutureHandler>;
+  private _iopubHandlers: Map<string, (msg: IKernelMsg) => void>;
+}
 
-    /**
-     * Handle an input message (execute_input).
-     *
-     * @function _handle_input message
-     */
-    _handle_input_message(msg): void {
-        var callbacks = this.get_callbacks_for_msg(msg.parent_header.msg_id);
-        if (!callbacks) {
-            // The message came from another client. Let the UI decide what to
-            // do with it.
-            events.trigger('received_unsolicited_message.Kernel', msg);
-        }
-    }
 
-    /**
-     * Dispatch IOPub messages to respective handlers. Each message
-     * type should have a handler.
-     *
-     * @function _handle_iopub_message
-     */
-    _handle_iopub_message(msg) : string {
-        var handler = this.get_iopub_handler(msg.header.msg_type);
-        if (handler !== undefined) {
-            return handler(msg);
-        }
-    }
-
-    /**
-     * @function _handle_input_request
-     */
-    _handle_input_request(request): void {
-        var header = request.header;
-        var content = request.content;
-        var metadata = request.metadata;
-        var msg_type = header.msg_type;
-        if (msg_type !== 'input_request') {
-            console.log("Invalid input request!", request);
-            return;
-        }
-        var callbacks = this.get_callbacks_for_msg(request.parent_header.msg_id);
-        if (callbacks) {
-            if (callbacks.input) {
-                callbacks.input(request);
-            }
-        }
-    }
+/**
+ * Bit flags for the Kernel future state.
+ */
+enum KernelFutureFlag {
+  GotReply = 0x1,
+  GotIdle = 0x2,
+  AutoDispose = 0x4,
+  IsDone = 0x8
+}
 
-    public id: string;
-    public name: string;
-    public kernel_service_url: string;
-    public kernel_url: string;
-    public ws_url: string;
-    public username: string;
-    public session_id: string;
-    public ws: WebSocket;
-    public info_reply: any;
-    public WebSocket: any;
-    public comm_manager: comm.CommManager;
-    public last_msg_id: string;
-    public last_msg_callbacks: any;
-    public reconnect_limit: number;
-
-    private _msg_callbacks: any;
-    private _msg_queue: Promise;
-    private _autorestart_attempt: number;
-    private _reconnect_attempt: number;
-    private _iopub_handlers: any;
 
+/**
+ * Implementation of a Kernel Future.
+ */
+class KernelFutureHandler extends Disposable implements IKernelFuture {
+
+  /**
+   * Dispose and unregister the future.
+   */
+  dispose(): void {
```

Comment:
dispose your  own state before calling `super.dispose()`. Once you call `super.dispose()` the object should be treated as deleted.